AdminServer/appscale/admin/__init__.py

```diff
@@ -0,0 +1,404 @@
+
+    try:
+      info_map = acc.get_app_info_map()
+    except AppControllerException as error:
+      logging.warning('Unable to fetch info map: {}'.format(error))
+      yield gen.sleep(1)
+      continue
+
+    try:
+      raise gen.Return(info_map[project_id]['nginx'])
+    except KeyError:
+      yield gen.sleep(1)
+      continue
+
+
+@gen.coroutine
+def wait_for_port_to_open(http_port, operation_id, deadline):
+  """ Waits until port is open.
+
+  Args:
+    http_port: An integer specifying the version's port number.
+    operation_id: A string specifying an operation ID.
+    deadline: A float containing a unix timestamp.
+  Raises:
+    OperationTimeout if the deadline is exceeded.
+  """
+  logging.debug('Waiting for {} to open'.format(http_port))
+  operation = operations[operation_id]
+
+  while True:
+    if time.time() > deadline:
+      message = 'Deploy operation took too long.'
+      operation['done'] = True
+      operation['error'] = {'message': message}
+      raise OperationTimeout(message)
+
+    sock = socket.socket()
+    result = sock.connect_ex((options.login_ip, http_port))
+    if result == 0:
+      break
+
+    yield gen.sleep(1)
+
+
+@gen.coroutine
+def wait_for_deploy(operation_id, acc):
+  """ Tracks the progress of a deployment.
+
+  Args:
+    operation_id: A string specifying the operation ID.
+    acc: An AppControllerClient instance.
+  """
+  operation = operations[operation_id]
+  start_time = time.time()
+  deadline = start_time + MAX_DEPLOY_TIME
+
+  http_port = yield wait_for_port_definition(operation_id, deadline, acc)
+  yield wait_for_port_to_open(http_port, operation_id, deadline)
+
+  operation['done'] = True
+  create_time = datetime.datetime.utcnow()
+  operation['response'] = utils.format_version(
+    operation, constants.ServingStatus.SERVING, create_time, http_port)
+  logging.info('Finished operation {}'.format(operation_id))
+
+
+class BaseHandler(web.RequestHandler):
+  """ A base handler. """
+  def authenticate(self):
+    """ Ensures requests are authenticated.
+
+    Raises:
+      CustomHTTPError if the secret is invalid.
+    """
+    if 'AppScale-Secret' not in self.request.headers:
+      message = 'A required header is missing: AppScale-Secret'
+      raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message=message)
+
+    if self.request.headers['AppScale-Secret'] != options.secret:
+      raise CustomHTTPError(HTTPCodes.UNAUTHORIZED, message='Invalid secret')
+
+  def write_error(self, status_code, **kwargs):
+    """ Writes a custom JSON-based error message.
+
+    Args:
+      status_code: An integer specifying the HTTP error code.
+    """
+    details = {'code': status_code}
+    if 'exc_info' in kwargs:
+      error = kwargs['exc_info'][1]
+      try:
+        details.update(error.kwargs)
+      except AttributeError:
+        pass
+
+    self.finish(json_encode({'error': details}))
+
+
+class VersionsHandler(BaseHandler):
+  """ Manages service versions. """
```

Comment:
No, it's version as in App Engine project services and versions. Though we only support one version (default) and one service (default), it makes sense to me to start using the structure and terminology of the API, which supports multiple services and modules.